### DISCUSSION TITLE
What is the format of clink_history?

### DISCUSSION THREAD

**Discussion by endolith** [chrisant996/clink#229] [Q&A] [answered]
I constantly get sync conflicts when I use a shared Cmder on two different computers, and I'd like to be able to merge the clink_history files, but I don't understand the format.

Lines like

```
conda install dot
conda install graphviz
```

become 
```
conda install dot
|onda install graphviz
```

While lines like 

```
|ytest
|onda update anaconda
|onda update --all
```

become 

```
|ytest
conda update anaconda
conda update --all
```

What does the pipe mean?  Why does it appear and disappear over time?

If I just keep the unpiped versions from each conflicted file, will it break anything?

**Comment by chrisant996** [comment, marked as answer]
- The first line (beginning with `|CTAG`) is important for preventing corruption.
- The rest of the lines are commands.
- When a line is deleted, its first character is replaced with a pipe character `|`.
- Periodically Clink will compact the file, by removing all deleted lines (i.e. beginning with a pipe character `|`) and removing old lines until the total number of lines is within the `history.max_lines` limit.

If you are seeing lines "lose" a pipe character at the beginning, then that is probably a result of the sync conflicts.  There's certainly not anything Clink ever does that can make a pipe character disappear (apart from deleting the entire line, when compacting the history file).

**Reply by endolith** [reply]
> When a line is deleted

How is a line deleted?

**Reply by chrisant996** [reply]
I recommend reading the Clink documentation on History, and then asking follow-up questions.

**Reply by endolith** [reply]
Meaning this? https://chrisant996.github.io/clink/clink.html#saved-command-history

So the pipes are being added in order to do "Clink reloads the master history list and prunes it not to exceed the history.max_lines setting."?

**Reply by endolith** [reply]
Is there some way to tell it to immediately clean out the lines that start with pipes?

**Reply by endolith** [reply]
Oh I found it on another site. 

Note to future self: 

1. Use Winmerge to combine the `clink_history.sync-conflict-…` files with the `clink_history` file.  Don't worry about lines that start with pipes.  
2. Run `clink history compact` to remove all the piped junk
3. Try not to use Cmder on both machines at the same time

**Comment by chrisant996** [comment]
The `history.dupe_mode` setting can delete old duplicates.
The `remove-history` command, while searching history (e.g. <kbd>Up</kbd>/<kbd>Down</kbd>), removes the current line from the history.
